cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -127,8 +128,11 @@ func NewCmdApply(apf *applyPlanFlags) *cobra.Command {
 	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be performed.")
 	cmd.Flags().BoolVar(&flags.etcdUpgrade, "etcd-upgrade", flags.etcdUpgrade, "Perform the upgrade of etcd.")
 	cmd.Flags().DurationVar(&flags.imagePullTimeout, "image-pull-timeout", flags.imagePullTimeout, "The maximum amount of time to wait for the control plane pods to be downloaded.")
-	// TODO: Register this flag in a generic place
-	cmd.Flags().StringVar(&flags.criSocket, "cri-socket", flags.criSocket, "Specify the CRI socket to connect to.")
+
+	// The CRI socket flag is deprecated here, since it should be taken from the NodeRegistrationOptions for the current
+	// node instead of the command line. This prevents errors by the users (such as attempts to use wrong CRI during upgrade).
+	cmdutil.AddCRISocketFlag(cmd.Flags(), &flags.criSocket)
```

Comment:
Does this mean there will be no way to force a specific CRI? In that case, what would happen if users have `docker` and `crio` running at the same time?

Reply:
i had concerns about this myself but forgot to mention at today's office hours. :\
maybe we still want a flag that is set to a value of "auto" by default.

Reply:
agree

Reply:
> In fact, strictly speaking, the only operations that should allow passing of CRI sockets should be init, join and reset.

do you think we should undeprecated the flag and have it with a value of "auto" by default?

Reply:
I don't think so, it's best that this flag is removed altogether. The only thing, that users can attempt to do with this flag is to change the CRI upon upgrade. However, I don't think, that this is going to work properly and I don't think, that we should support such user story.

Reply:
but how would we handle the scenario @inercia mentioned:
> Does this mean there will be no way to force a specific CRI? In that case, what would happen if users have docker and crio running at the same time?

i.e. multiple CRIs installed. how would the user pick one of the sockets?

Reply:
The persisted `NodeRegistrationOptions` in the cluster for this node is going to contain a CRI socket, that was detected upon init/join. Hence, no detection will be required.